internal/httputil/ratelimiter.go

```diff
@@ -0,0 +1,62 @@
+package httputil
+
+import (
+	"net/http"
+	"sync"
+
+	"golang.org/x/time/rate"
+)
+
+// RateLimiter wraps the provided RoundTripper with a limiter allowing 10
+// requests/second/host.
+//
+// It responds to HTTP 429 responses by automatically decreasing the rate.
+func RateLimiter(next http.RoundTripper) http.RoundTripper {
+	return &ratelimiter{
+		rt: next,
+		lm: sync.Map{},
+	}
+}
+
+// Ratelimiter implements the limiting by using a concurrent map and Limiter
+// structs.
+type ratelimiter struct {
+	rt http.RoundTripper
+	lm sync.Map
+}
+
+const lim = 10
+
+// RoundTrip implements http.RoundTripper.
+func (r *ratelimiter) RoundTrip(req *http.Request) (*http.Response, error) {
+	key := req.URL.Host
+	li, ok := r.lm.Load(key)
+	if !ok {
+		// Limiter allows "lim" per sec, one at a time.
+		l := rate.NewLimiter(rate.Limit(lim), 1)
+		li, _ = r.lm.LoadOrStore(key, l)
+	}
+	l := li.(*rate.Limiter)
+	if err := l.Wait(req.Context()); err != nil {
+		return nil, err
+	}
+	res, err := r.rt.RoundTrip(req)
+	// This seems to be the contract that http.Transport implements.
+	if err != nil {
+		return nil, err
+	}
+	// This doesn't seem entirely correct, but will at least make number go
+	// down.
+	if res.StatusCode == http.StatusTooManyRequests {
+		l.SetLimit(detune(l.Limit()))
+	}
```

Comment:
I may be reading this wrong, but when/if the downstream recovers, would the limit be "stuck" at a lower than optimal rate?

Reply:
Yes.

I'm not sure what to do to re-ramp the limit. Using `SetLimitAt` would result in the limit getting similarly stuck, by my thinking. Perhaps a second Limiter with token representing an increase to the limit?

Reply:
I am not so familiar with the request patterns or how a certain node can become overwhelmed, but would it make sense to have a second case (or an else case) that increases the Limit (retune) when `status == <good status>` and then taking the `Min(newLimit, lim)` to ensure you don't go over the max? (In general, detune should reduce faster than retune increases it).

In the long run it seems like using a client to mitigate against server errors would lead to a different loadbalancing strategy to the `RoundTrip`, but that seems like a discussion for a different time.